### PR TITLE
KAFKA-5128: check inter broker in transactional methods

### PR DESCRIPTION
**Pull request by dguy**
Add check in `KafkaApis` that the inter broker protocol version is at least `KAFKA_0_11_0_IV0`, i.e., supporting transactions

### CODE REVIEW

**Comment by dguy** [comment]
I added a check for inter broker protocol version in all the transactional APIs as it seems to make sense to bail out early if the version is not at least what is expected.
Additionally in `handleAddPartitionToTxnRequest` check if all the topics in the request are using the correct message format. error out if any aren't.

Still need to add tests, but this is more to make sure i'm thinking about this correctly.

@guozhangwang @hachikuji @apurvam @ijuma

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 1]
```diff
@@ -117,11 +117,28 @@ class KafkaApis(val requestChannel: RequestChannel,
         case ApiKeys.DELETE_RECORDS => handleDeleteRecordsRequest(request)
         case ApiKeys.INIT_PRODUCER_ID => handleInitProducerIdRequest(request)
         case ApiKeys.OFFSET_FOR_LEADER_EPOCH => handleOffsetForLeaderEpochRequest(request)
-        case ApiKeys.ADD_PARTITIONS_TO_TXN => handleAddPartitionToTxnRequest(request)
-        case ApiKeys.ADD_OFFSETS_TO_TXN => handleAddOffsetsToTxnRequest(request)
-        case ApiKeys.END_TXN => handleEndTxnRequest(request)
-        case ApiKeys.WRITE_TXN_MARKERS => handleWriteTxnMarkersRequest(request)
-        case ApiKeys.TXN_OFFSET_COMMIT => handleTxnOffsetCommitRequest(request)
+        case ApiKeys.ADD_PARTITIONS_TO_TXN => handleWithInterBrokerVersionAtLeast(KAFKA_0_11_0_IV0, request, (error:Errors) => {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: I think we could turn this into a simple assertion which could be added to the respective handlers so we don't make this block messy. Something like this:

```scala
def ensureInterBrokerVersion(version: Short) {
  if (config.interBrokerProtocolVersion < version)
    throw new InvalidRequestException()
}
```
The exception will be caught and passed to `getErrorResponse`, so the effect should be the same.

Also, I'm debating whether InvalidRequest is what we want. We could use UnsupportedVersion also, but it's a bit odd that we raise this after we have already advertised support in the ApiVersions response. Perhaps that means we shouldn't even advertise them in ApiVersions until the inter-broker protocol has been updated?

cc @ijuma

Comment (ijuma) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: I think both are good ideas. It's a new use of `inter.broker.protocol.version`, but it seems fair to exclude an API that will always throw an error.

Comment (dguy) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: I've done the first part, but i don't get what i'm supposed to do for the second part? We have `ApiVersion` which is what i'm using to determine if we have `inter.broker.protocol.version` set to at least `KAFKA_0_11_0_IV0`. We then have `ApiVersionsResponse` that uses `ApiKeys.values()` to build a list of an inner class `ApiVersion`
How do we tie an `ApiVersion` as in `KAFKA_0_11_0_IV0` to the list of supported operations?

Comment (ijuma) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: `ApiVersionsResponse.apiVersionsResponse` (called from `KafkaApis`) would have to be changed to receive the valid `ApiKeys`.

Comment (dguy) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Right, but how do we determine the valid `ApiKeys` for a particular `ApiVersion`? It must be staring me in the face somewhere!

Comment (hachikuji) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: @dguy I think there are a couple ways to do it. If you look in `ApiVersionsResponse`, we have a static response  `API_VERSIONS_RESPONSE` which contains all of the APIs that the broker supports (without regard to inter broker protocol). One option might be to turn that into a map which is keyed by the inter broker protocol. Maybe the minimum inter-broker protocol could then be added as a field to `ApiKeys`?

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 2]
```diff
@@ -1552,30 +1560,45 @@ class KafkaApis(val requestChannel: RequestChannel,
...
       } else {
-        // Send response callback
-        def sendResponseCallback(error: Errors): Unit = {
-          def createResponse(requestThrottleMs: Int): AbstractResponse = {
-            val responseBody: AddPartitionsToTxnResponse = new AddPartitionsToTxnResponse(requestThrottleMs,
-              partitionsToAdd.asScala.map{tp => (tp, error)}.toMap.asJava)
-            trace(s"Completed $transactionalId's AddPartitionsToTxnRequest with partitions $partitionsToAdd: errors: $error from client ${request.header.clientId}")
-            responseBody
+        val (_, incorrectMessageFormat) = partitionsToAdd.asScala.partition { tp =>
+          replicaManager.getMagic(tp) match {
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: Hmm... I think this only works if the partition is hosted on the same broker.

Comment (dguy) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: ok - so how would we check this?

Comment (ijuma) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: The config lives in ZK. We have a listener that keeps the LogConfig in each Log up to date. The ReplicaManager  gets the config from the Log, which only works if the broker is hosting that partition.

The simple solution is to fetch from ZK, but it's not ideal to do it for every request. An alternative is to keep a cache for every topic. We can probably reuse the existing TopicConfigHandler to keep the cache updated.

Comment (dguy) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Thanks @ijuma

Comment (hachikuji) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Is the check really necessary? The other option is to skip it and handle the UNSUPPORTED_FOR_MESSAGE_FORMAT error from the WriteTxnMarker response. I think we will probably have to handle that error anyway, so maybe we don't need the additional complexity of config caching and such.

Comment (dguy) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: @hachikuji doesn't that seem a bit late? Why would we not error out as soon as you try to add a partition that has an unsupported format, i.e., error out as early as possible?

Comment (hachikuji) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: If we had an easy way to check for unsupported format, then I agree, but if we need to introduce a new path for accessing non-local configs just for this use case, I'd say skip it. Instead of getting the error here, the user will get one when they try to write to the partition and will have to abort the transaction. As long as we expect the error from WriteTxnMarker, no harm done. What do you think?

Comment (dguy) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Makes it simpler, so ok.

Comment (dguy) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: @hachikuji on second thoughts. I think there is one problem with your suggested approach. By the time we send the WriteTxnMarkers request from TC we have already responded to the client with Errors.NONE. So the client would think the commit was successful.

Comment (hachikuji) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: My assumption is that if the topic doesn't support the message format, then the producer will be doomed to abort as soon as they try to write to it. If they go ahead and commit anyway (maybe because of a client bug), I still think it's "ok" in the sense that no transactional data could have been written to the topic which didn't support the message format, so there will be no need to write any markers (committing is a no-op in some sense). This is similar to the problem with topic deletion, which could occur after we have accepted a commit.

Comment (dguy) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Ok - i believe you :-)

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 1 above)
Comment (dguy) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Should all `ApiKey` really be tied to the inter broker protocol version? This is the first time in any request it has been checked. So that says to me that for most requests the inter broker protocol version is irrelevant. So is limiting the `ApiKeys` returned in `ApiVersionsResponse` based on inter broker protocol actually a valid thing to do?

Assuming that it is valid... What are we adding to ApiKeys to indicate the inter broker protocol version? `ApiVersion.id` ? `ApiVersion.messageFormatVersion`

Comment (ijuma) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: `ApiKeys` could have a `minMagic` or `minMessageFormatVersion` field. By default there is no minimum (which is the common case). For the ones that have a minimum, they can use an overloaded constructor to indicate that. Maybe we should change the constructor that takes `clusterAction` to have an additional parameter and add a constructor that takes id and name.

Comment (dguy) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Thanks @ijuma seems like it would work

File: clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java [hunk 3]
```diff
@@ -19,32 +19,33 @@
...
-    SYNC_GROUP(14, "SyncGroup", false),
-    DESCRIBE_GROUPS(15, "DescribeGroups", false),
-    LIST_GROUPS(16, "ListGroups", false),
-    SASL_HANDSHAKE(17, "SaslHandshake", false),
-    API_VERSIONS(18, "ApiVersions", false) {
+    PRODUCE(0, "Produce"),
+    FETCH(1, "Fetch"),
+    LIST_OFFSETS(2, "Offsets"),
+    METADATA(3, "Metadata"),
+    LEADER_AND_ISR(4, "LeaderAndIsr", true, RecordBatch.MAGIC_VALUE_V0),
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java]: nit: maybe we could just add a third constructor?

File: clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java [hunk 4]
```diff
@@ -93,12 +94,19 @@ public Struct parseResponse(short version, ByteBuffer buffer) {
     /** indicates if this is a ClusterAction request used only by brokers */
     public final boolean clusterAction;
 
-    ApiKeys(int id, String name, boolean clusterAction) {
+    public final byte minMagic;
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java]: I think a couple comments on how this is used would be helpful. It's not obvious that we use it as an indirection to get to the inter-broker protocol version. Maybe a name like`minRequiredInterBrokerMagic` or something like that would make it a little clearer?

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 5]
```diff
@@ -1572,27 +1580,30 @@ class KafkaApis(val requestChannel: RequestChannel,
         sendResponseMaybeThrottle(request, requestThrottleMs =>
           new AddPartitionsToTxnResponse(requestThrottleMs, partitionErrors.asJava))
       } else {
-        def sendResponseCallback(error: Errors): Unit = {
-          def createResponse(requestThrottleMs: Int): AbstractResponse = {
-            val responseBody: AddPartitionsToTxnResponse = new AddPartitionsToTxnResponse(requestThrottleMs,
-              partitionsToAdd.asScala.map{tp => (tp, error)}.toMap.asJava)
-            trace(s"Completed $transactionalId's AddPartitionsToTxnRequest with partitions $partitionsToAdd: errors: $error from client ${request.header.clientId}")
-            responseBody
+          // Send response callback
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: Seems we can revert these changes.

**Review by hachikuji — COMMENTED** [review]
Thanks for the patch. Just a couple minor comments.

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks for the patch!
